routes/artifacts.js

```diff
@@ -0,0 +1,597 @@
+var assert  = require('assert');
+var Promise = require('promise');
+
+// Common schema prefix
+const SCHEMA_PREFIX_CONST = 'http://schemas.taskcluster.net/queue/v1/';
+
+// Maximum number of artifacts to list
+const MAX_ARTIFACTS_LISTING = 10 * 1000;
+
+/** Post artifact */
+api.declare({
+  method:     'post',
+  route:      '/task/:taskId/runs/:runId/artifacts/:name(*)',
+  name:       'createArtifact',
+  scopes: [
+    [
+      'queue:create-artifact:<name>',
+      'assume:worker-id:<workerGroup>/<workerId>'
+    ]
+  ],
+  deferAuth:  true,
+  input:      SCHEMA_PREFIX_CONST + 'post-artifact-request.json#',
+  output:     SCHEMA_PREFIX_CONST + 'post-artifact-response.json#',
+  title:      "Create Artifact",
+  description: [
+    "This API end-point creates an artifact for a specific run of a task. This",
+    "should **only** be used by a worker currently operating on this task, or",
+    "from a process running within the task (ie. on the worker).",
+    "",
+    "All artifacts must specify when they `expires`, the queue will",
+    "automatically take care of deleting artifacts past their",
+    "expiration point. This features makes it feasible to upload large",
+    "intermediate artifacts from data processing applications, as the",
+    "artifacts can be set to expire a few days later.",
+    "",
+    "We currently support 4 different `storageType`s, each storage type have",
+    "slightly different features and in some cases difference semantics.",
+    "",
+    "**S3 artifacts**, is useful for static files which will be stored on S3.",
+    "When creating an S3 artifact is create the queue will return a pre-signed",
+    "URL to which you can do a `PUT` request to upload your artifact. Note",
+    "that `PUT` request **must** specify the `content-length` header and",
+    "**must** give the `content-type` header the same value as in the request",
+    "to `createArtifact`.",
+    "",
+    "**Azure artifacts**, are stored in _Azure Blob Storage_ service, which",
+    "given the consistency guarantees and API interface offered by Azure is",
+    "more suitable for artifacts that will be modified during the execution",
+    "of the task. For example docker-worker has a feature that persists the",
+    "task log to Azure Blob Storage every few seconds creating a somewhat",
+    "live log. A request to create an Azure artifact will return a URL",
+    "featuring a [Shared-Access-Signature]" +
+    "(http://msdn.microsoft.com/en-us/library/azure/dn140256.aspx),",
+    "refer to MSDN for further information on how to use these.",
+    "",
+    "**Reference artifacts**, only consists of meta-data which the queue will",
+    "store for you. These artifacts really only have a `url` property and",
+    "when the artifact is requested the client will be redirect the URL",
+    "provided with a `303` (See Other) redirect. Please note that we cannot",
+    "delete artifacts you upload to other service, we can only delete the",
```

Comment:
Maybe we should include this warning in the description of the other services as well.  Duplication, but if someone's only looking for azure info, they'll likely only look at the azure section